docs/architecture/adr-006-subspace-module.md

```diff
@@ -0,0 +1,159 @@
+# ADR 006: Subspace module
+
+## Changelog
+
+- December 15th, 2021: Initial draft
+
+## Status
+
+DRAFT
+
+## Abstract
+This ADR defines the `x/subspaces` module which allows users to create and manage the representation of different social networks inside which contents will be created.  
+
+## Context
+In order to support building independent social networks, it is vital for Desmos to expose a mechanism that allows to replicate as much as possible the current stace of social networks in which each platform has different economic model, set of administrators, Terms of Services and ways of enforcing such terms.  
+
+## Decision
+We will create a module named `subspaces` which provided users the ability to create independent spaces inside Desmos, each one having its own administrators, ToS and tokenomics.
+
+### Types 
+Subspaces must always have an account that is elected as the _owner_ and should act as the final decision maker and accountable user. Additionally, each subspace can have the following data:
+
+* a _human-readable name_ and a _description_, in order to allow users to easily identify the scope of such subspace;
+* a set of _connections_ with external applications, allowing users to verify the validity of such subspace and avoid any possible fakes;
+* an _ACL_ (access-control list) defining the permissions that different kind of users have inside the subspace itself (e.g. administrators, moderators, etc.).
+
+#### Subspace
+
+```go
+// Subspace contains all the data related to a Desmos subspace
+type Subspace struct {
+  // Unique identifier of this subspace
+  ID uint64
+
+  // Human-readable name of the subspace
+  Name string
+  
+  // Optional description of this subspace
+  Description string
+  
+  // Address of the user that owns the subspace
+  Owner  string
+
+  // Address of the subspace creator
+  Creator string
+  
+  // Represents the account that is associated with the subspace and
+  // should be used to connect external applications to verify this subspace
+  Treasury string
+
+  // the creation time of the subspace
+  CreationTime time.Time
+}
+```
+
+#### ACL
+In order to easily implement an ACL, we will use a simple set of keys made as follows: 
+```
+ACLPrefix + User Address -> ACL Value
```

Comment:
I guess it should be `ACLPrefix + SubspaceID + User Address` since the ACL is based on the subspace.

Reply:
@dadamu Yes you are right. Otherwise without it once we cannot properly iterate over the mods/admin of a particular subspace.
Maybe @RiccardoM you implicit thought that the `subspaceID` is included in the ACLPrefix?
I suggest to have something like:
* `ACLSubspacePrefix` = `ACLPrefix + Subspace` which will make it easy to iterate over all the users of a subspace.
* `ACLUserKey` = `ACLSubspacePrefix + User Address` to check on a specific user ACL value